### PR TITLE
Fix RBAC with compliance and admission controller

### PR DESCRIPTION
**Pull request by clamoriniere**
### What does this PR do?

Fix missing Operator RBAC permission when the DCA is activated:
* Create DCA RBAC even if the ExternalMetrics service is not enabled
* Add missing DCA RBAC for Admission-webhook 
* Add missing DCA RBAC compliance
* Add missing DCA RBAC for orchestrator app: cluster-id detection.

### Motivation

* Solve several permission error in the Operator for the creation of the DCA RBAC
* Fix DCA feature due to missing RBAC.

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Use the `examples/datadog-agent-all.yaml` and verify that the Operator and the DCA pod logs don't contain RBAC error.

To test the DCA deployment without ExternalMetric, you can change the ClusterAgent configuration by:
```yaml
  clusterAgent:
    image:
      name: "datadog/cluster-agent:latest"
    config:
      externalMetrics:
        enabled: false
```

### CODE REVIEW

File: chart/datadog-operator/templates/clusterrole.yaml [hunk 1]
```diff
@@ -57,4 +72,166 @@ rules:
...
+  - admissionregistration.k8s.io
+  resources:
+  - mutatingwebhookconfigurations
+  - secrets
+  verbs:
+  - get
+  - list
+  - watch
+  - create
+  - update
```
Comment (ahmed-mez) [inline chart/datadog-operator/templates/clusterrole.yaml:163]: why do we define these twice? we have them defined above

Comment (clamoriniere) [reply chart/datadog-operator/templates/clusterrole.yaml:163]: Also the "secrets" here look wrong.

Comment (clamoriniere) [reply chart/datadog-operator/templates/clusterrole.yaml:163]: wrong cut/past

Comment (ahmed-mez) [reply chart/datadog-operator/templates/clusterrole.yaml]: same thing in deploy/clusterrole

File: examples/datadog-agent-all.yaml [hunk 2]
```diff
@@ -18,3 +18,16 @@ spec:
     systemProbe:
       enabled: true
       bpfDebugEnabled: true
+    security:
```
Comment (xornivore) [inline examples/datadog-agent-all.yaml:21]: 👍

**Review by xornivore — APPROVED** [review]

File: chart/datadog-operator/templates/clusterrole.yaml [hunk 3]
```diff
@@ -57,4 +61,165 @@ rules:
...
   - "get"
   - "watch"
+- apiGroups:
+  - autoscaling
+  resources:
+  - horizontalpodautoscalers
+  verbs:
+  - list
+  - watch
+## To fix: should be role not a clusterrole
```
Comment (clamoriniere) [inline chart/datadog-operator/templates/clusterrole.yaml:71]: All those configmap are only accessed by the DCA and should be in the DCA namespace. So I think we can move it in the role.

File: examples/datadog-agent-all.yaml [hunk 4]
```diff
@@ -18,3 +18,16 @@ spec:
...
+      runtime:
+        enabled: false
+  clusterAgent:
+    image:
+      name: "datadog/cluster-agent:latest"
+    config:
+      externalMetrics:
+        enabled: true
+      admissionController:
+        enabled: true
```
Comment (xornivore) [inline examples/datadog-agent-all.yaml:33]: nit: new line missing

File: pkg/controller/datadogagent/clusteragent.go [hunk 5]
```diff
@@ -622,18 +622,25 @@ func (r *ReconcileDatadogAgent) manageClusterAgentRBACs(logger logr.Logger, dda
...
 	// Create or update ClusterRole
 	clusterRole := &rbacv1.ClusterRole{}
 	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: rbacResourcesName}, clusterRole); err != nil {
 		if errors.IsNotFound(err) {
 			return r.createClusterAgentClusterRole(logger, dda, rbacResourcesName, clusterAgentVersion)
 		}
 		return reconcile.Result{}, err
-	}
-	if result, err := r.updateIfNeededClusterAgentClusterRole(logger, dda, rbacResourcesName, clusterAgentVersion, clusterRole); err != nil {
+	} else if result, err := r.updateIfNeededClusterAgentClusterRole(logger, dda, rbacResourcesName, clusterAgentVersion, clusterRole); err != nil {
```
Comment (xornivore) [inline pkg/controller/datadogagent/clusteragent.go:641]: Is this change functional or cosmetic here (joining the two checks into `else if`)? Personally I prefer to avoid else if when the conditions are complex like this and there are error checks involved - it makes it rather hard to follow and may make the code susceptible to misinterpretation and refactoring leading to bugs.

File: pkg/controller/datadogagent/clusteragent.go [hunk 6]
```diff
@@ -646,7 +653,29 @@ func (r *ReconcileDatadogAgent) manageClusterAgentRBACs(logger logr.Logger, dda
...
 	}
-
+	// Create or update Role
+	role := &rbacv1.Role{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: rbacResourcesName, Namespace: dda.Namespace}, role); err != nil {
+		if errors.IsNotFound(err) {
+			return r.createClusterAgentRole(logger, dda, rbacResourcesName, clusterAgentVersion)
+		}
+		return reconcile.Result{}, err
+	} else if result, err := r.updateIfNeededClusterAgentRole(logger, dda, rbacResourcesName, clusterAgentVersion, role); err != nil {
```
Comment (xornivore) [inline pkg/controller/datadogagent/clusteragent.go:663]: Similar to the above comment about `else if`.

File: pkg/controller/datadogagent/clusteragent.go [hunk 7]
```diff
@@ -1070,25 +1075,25 @@ func buildClusterAgentClusterRole(dda *datadoghqv1alpha1.DatadogAgent, name, age
 				datadoghqv1alpha1.UpdateVerb},
 		})
 
-		// Replicasets
+		// ExtendedDaemonsetReplicaSets
```
Comment (ahmed-mez) [inline pkg/controller/datadogagent/clusteragent.go:1078]: do we need this?

Comment (clamoriniere) [reply pkg/controller/datadogagent/clusteragent.go:1085]: Yes for pod, managed by an extendeddaemonset

Comment (clamoriniere) [reply pkg/controller/datadogagent/clusteragent.go:1085]: it is to be futur proof :)

File: pkg/controller/datadogagent/clusteragent.go (hunk 5 above)
Comment (clamoriniere) [reply pkg/controller/datadogagent/clusteragent.go]: agree, I have reverted this change

**Review by vboulineau — APPROVED** [review]
